crypto/bls/signature_batch.go

```diff
@@ -54,3 +54,86 @@ func (s *SignatureBatch) Copy() *SignatureBatch {
 		Messages:   messages,
 	}
 }
+
+// RemoveDuplicates removes duplicate signature sets from the signature batch.
+func (s *SignatureBatch) RemoveDuplicates() (int, *SignatureBatch) {
+	if len(s.Signatures) == 0 || len(s.PublicKeys) == 0 || len(s.Messages) == 0 {
+		return 0, s
+	}
+	sigMap := make(map[string]int)
+	duplicateSet := make(map[int]bool)
+	for i := 0; i < len(s.Signatures); i++ {
+		currSig := string(s.Signatures[i])
+		if sigIdx, ok := sigMap[currSig]; ok {
+			if s.PublicKeys[sigIdx].Equals(s.PublicKeys[i]) &&
+				s.Messages[sigIdx] == s.Messages[i] {
+				duplicateSet[i] = true
+				continue
+			}
+		}
+		sigMap[currSig] = i
+	}
+
+	sigs := s.Signatures[:0]
+	pubs := s.PublicKeys[:0]
+	msgs := s.Messages[:0]
+
+	for i := 0; i < len(s.Signatures); i++ {
+		if duplicateSet[i] {
+			continue
+		}
+		sigs = append(sigs, s.Signatures[i])
+		pubs = append(pubs, s.PublicKeys[i])
+		msgs = append(msgs, s.Messages[i])
+	}
+
+	s.Signatures = sigs
+	s.PublicKeys = pubs
+	s.Messages = msgs
+
+	return len(duplicateSet), s
+}
+
+// AggregateBatch aggregates common messages in the provided batch to
+// reduce the number of pairings required when we finally verify the
+// whole batch.
+func (s *SignatureBatch) AggregateBatch() (*SignatureBatch, error) {
+	if len(s.Signatures) == 0 || len(s.PublicKeys) == 0 || len(s.Messages) == 0 {
+		return s, nil
+	}
+	msgMap := make(map[[32]byte]*SignatureBatch)
+
+	for i := 0; i < len(s.Messages); i++ {
+		currMsg := s.Messages[i]
+		currBatch, ok := msgMap[currMsg]
+		if ok {
+			currBatch.Signatures = append(currBatch.Signatures, s.Signatures[i])
```

Comment:
We should add a defensive check for length of public keys, messages, and signatures being the same, otherwise a fuzzer will easily make this panic